lib/rolify/adapters/active_record/role_adapter.rb

```diff
@@ -9,14 +9,20 @@ def where(relation, *args)
       end
 
       def where_strict(relation, args)
-        return relation.where(:name => args[:name]) if args[:resource].blank?
-        resource = if args[:resource].is_a?(Class)
-                     {class: args[:resource].to_s, id: nil}
-                   else
-                     {class: args[:resource].class.name, id: args[:resource].id}
-                   end
-
-        relation.where(:name => args[:name], :resource_type => resource[:class], :resource_id => resource[:id])
+        wrap_conditions = relation.name != role_class.name
+        processed_args = { :name => args[:name] }
+        conditions = wrap_conditions ? { role_table => processed_args } : processed_args
+        return relation.where(conditions) if args[:resource].blank?
+
+        if args[:resource].is_a?(Class)
+          processed_args[:resource_type] = args[:resource].to_s
```

Comment:
And apparently I didn't actually submit my comment... Would refactor this code to avoid mutating processed_args w/reliance on it being referenced within the conditions hash. It's a bit tough because of the wrapping logic, not sure I have a magic proposal at the moment

Reply:
I tried a few approaches and the mutating was probably the simplest but I understand why it might be objectionable. I'll take another look in the coming days to see if I can reason something better.

Reply:
Yeah - my only suggestion without moving too much of this code around could be doing the wrapping at the return sites? I think it would be more legible at the cost of being a little more repetitive.

Reply:
OK I've updated things to avoid the nested hash mutating. There's still a tiny mutation in there but I think it's a bit cleaner. What do you think?